nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,570 @@
+  cmd_vel.twist.angular.z = angular_vel;
+  return cmd_vel;
+}
+
+bool RegulatedPurePursuitController::shouldRotateToPath(
+  const geometry_msgs::msg::PoseStamped & carrot_pose, double & angle_to_path)
+{
+  // Whether we should rotate robot to rough path heading
+  angle_to_path = atan2(carrot_pose.pose.position.y, carrot_pose.pose.position.x);
+  return use_rotate_to_heading_ && fabs(angle_to_path) > rotate_to_heading_min_angle_;
+}
+
+void RegulatedPurePursuitController::rotateToHeading(
+  double & linear_vel, double & angular_vel,
+  const double & angle_to_path, const geometry_msgs::msg::Twist & curr_speed)
+{
+  // Rotate in place using max angular velocity / acceleration possible
+  linear_vel = 0.0;
+  const double sign = angle_to_path > 0.0 ? 1.0 : -1.0;
+  angular_vel = sign * rotate_to_heading_angular_vel_;
+
+  const double & dt = control_duration_;
+  const double min_feasible_angular_speed = curr_speed.angular.z - max_angular_accel_ * dt;
+  const double max_feasible_angular_speed = curr_speed.angular.z + max_angular_accel_ * dt;
+  angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
+}
+
+geometry_msgs::msg::PoseStamped RegulatedPurePursuitController::getLookAheadPoint(
+  const double & lookahead_dist,
+  const nav_msgs::msg::Path & transformed_plan)
+{
+  // Find the first pose which is at a distance greater than the lookahead distance
+  auto goal_pose_it = std::find_if(
+    transformed_plan.poses.begin(), transformed_plan.poses.end(), [&](const auto & ps) {
+      return hypot(ps.pose.position.x, ps.pose.position.y) >= lookahead_dist;
+    });
+
+  // If the no pose is not far enough, take the last pose
+  if (goal_pose_it == transformed_plan.poses.end()) {
+    goal_pose_it = std::prev(transformed_plan.poses.end());
+  }
+
+  return *goal_pose_it;
+}
+
+bool RegulatedPurePursuitController::isCollisionImminent(
+  const geometry_msgs::msg::PoseStamped & robot_pose,
+  const geometry_msgs::msg::PoseStamped & carrot_pose,
+  const double & curvature, const double & linear_vel,
+  const double & angular_vel)
+{
+  // Note(stevemacenski): This may be a bit unusual, but the robot_pose is in
+  // odom frame and the carrot_pose is in robot base frame. We need to collision
+  // check in odom frame, so all values will be relative to robot base pose.
+  // But we can still use the carrot pose in odom to find various quantities.
+
+  geometry_msgs::msg::PoseStamped carrot_in_odom;
+  if (!transformPose(costmap_ros_->getGlobalFrameID(), carrot_pose, carrot_in_odom))
+  {
+    RCLCPP_ERROR(logger_, "Unable to get carrot pose in odom frame, failing collision check!");
+    return true;
+  }
+
+  // check current point and carrot point are OK, most likely ones to be in collision
+  if (inCollision(carrot_in_odom.pose.position.x, carrot_in_odom.pose.position.y) ||
+    inCollision(robot_pose.pose.position.x, robot_pose.pose.position.y))
+  {
+    return true;
+  }
+
+  // debug messages
```

Comment:
Looks like this is lookahead arc path made for debugging visualization. This might be a confusion about `RCLCPP_DEBUG()` which are also called "debug messages". I think, it is better to change this comment to something more descriptive.